query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala

```diff
@@ -395,75 +397,281 @@ final case class LabelCardinalityExec(queryContext: QueryContext,
     s" startMs=$startMs, endMs=$endMs"
 }
 
-final case object TopkCardExec {
+/**
+ * Contains utilities for all TsCardinality materialize() derivatives.
+ */
+final case object TsCardExec {
   // TODO: tune this
-  // See here for help choosing a size:
-  //   https://datasketches.apache.org/docs/Frequency/FrequentItemsErrorTable.html
-  val MAX_ITEMSKETCH_MAP_SIZE = 256;
+  val MAX_RESPONSE_SIZE = 5000
+
+  // just to keep memStore.topKCardinality calls consistent
+  val ADD_INACTIVE = true
+
+  val OVERFLOW_NAME = "_overflow_".utf8
+
+  case class CardCounts(active: Int, total: Int) {
+    require(total >= active, "total must be at least as large as active")
+  }
+
+  // --- Serialization Utilities are Below ---
+
+  val NUM_INT_BYTES: Int = 4
+
+  private def intToBytes(num: Int, base: Array[Byte], offset: Int): Unit = {
+    val bytes = java.nio.ByteBuffer.wrap(base, offset, NUM_INT_BYTES)
+    bytes.putInt(num)
+  }
+
+  private def bytesToInt(base: Array[Byte], offset: Int): Int = {
+    val reader = java.nio.ByteBuffer.wrap(base, offset, NUM_INT_BYTES)
+    reader.getInt()
+  }
+
+  def deSerializeMap(bytes: Array[Byte]): Map[Seq[ZeroCopyUTF8String], CardCounts] = {
+    val result = new mutable.HashMap[Seq[ZeroCopyUTF8String], CardCounts]
+    val prefixSize = bytesToInt(bytes, 0)
+    var ibyte = NUM_INT_BYTES;  // since we just read prefixSize
+    while (ibyte < bytes.length) {
+      // deser the prefix first
+      val prefix = new mutable.ArraySeq[ZeroCopyUTF8String](prefixSize)
+      for (iname <- 0 until prefixSize) {
+        val size = bytesToInt(bytes, ibyte)
+        ibyte = ibyte + NUM_INT_BYTES
+        val name = ZeroCopyUTF8String(bytes, ibyte, size)
+        ibyte = ibyte + size
+        prefix(iname) = name
+      }
+      // now deser the counts
+      val active = bytesToInt(bytes, ibyte)
+      ibyte = ibyte + NUM_INT_BYTES
+      val total = bytesToInt(bytes, ibyte)
+      ibyte = ibyte + NUM_INT_BYTES
+      result.update(prefix, CardCounts(active, total))
+    }
+    result.toMap
+  }
+
+  def serializeMap(myMap: Map[Seq[ZeroCopyUTF8String], CardCounts]): Array[Byte] = {
+    // Need a quick way to access the underlying array.
+    val builder = new ArrayBuilder.ofByte
+
+    // Pre-allocate a space to store the bytes for use with a
+    //   Java ByteBuffer. Then we'll addAll() to the ArrayBuilder.
+    // TODO: unsafe casts to bypass individual byte writes?
```

Comment:
This would probably warrant a new `ArrayBuffer`-esque class with `appendSize` and `appendBlob` methods, and that feels like it would be overkill.

Reply:
Resolved after schema change